### PR TITLE
[6/8] ENH: update the pixel spot detector to use IntensityTable and Codebook

### PR DESCRIPTION
**Pull request by ambrosejcarr**
- Updates pixel_spot_detector pipeline component to use IntensityTable and Codebook objects.

### CODE REVIEW

File: starfish/pipeline/features/pixels/pixel_spot_detector.py [hunk 1]
```diff
@@ -1,27 +1,21 @@
...
 class PixelSpotDetector(PixelFinderAlgorithmBase):
 
-    _MerfishDecoderResults = collections.namedtuple(
-        '_MerfishDecoderResults',
-        ['result', 'decoded_img', 'label_img', 'spot_props'])
-
     def __init__(
-            self, codebook: str, distance_threshold: float=0.5176,
-            magnitude_threshold: int=1, area_threshold: int=2, crop_size: int=40, **kwargs) -> None:
+            self, codebook: Union[str, Codebook], distance_threshold: float=0.5176,
```
Comment (ttung) [inline starfish/pipeline/features/pixels/pixel_spot_detector.py]: Would prefer that the constructor is more specific about what it accepts (probably Codebook?) and provide factory methods to call the constructor.

File: starfish/pipeline/features/pixels/pixel_spot_detector.py [hunk 2]
```diff
@@ -34,132 +28,31 @@ def __init__(
...
-
-        res = pd.DataFrame(mat.T)
-        res['spot_id'] = range(len(res))
-        res = melt(
-            df=res,
-            new_index_name='barcode_index',
-            new_value_name='intensity',
-            melt_columns=range(num_bits)
+    def find(self, stack: ImageStack) \
+            -> Tuple[IntensityTable, List, np.ndarray, np.ndarray, np.ndarray]:
```
Comment (ttung) [inline starfish/pipeline/features/pixels/pixel_spot_detector.py]: would prefer a namedtuple or a data type.

File: starfish/pipeline/features/pixels/pixel_spot_detector.py [hunk 3]
```diff
@@ -34,132 +28,31 @@ def __init__(
             spots with intensity less than this value are filtered (default 1)
         area_threshold : int
             spots with total area less than this value are filtered
-        crop_size : int
-            ??? # TODO ambrosejcarr what is this?
+        crop_size : Tuple[int, int, int]
```
Comment (ttung) [inline starfish/pipeline/features/pixels/pixel_spot_detector.py]: would prefer explicit crop parameters.

File: starfish/intensity_table.py [hunk 4]
```diff
@@ -276,3 +278,138 @@ def synthetic_intensities(
...
+        gene_to_int = {v: k for (k, v) in int_to_gene.items()}
+
+        # map genes to ints
+        gene_list = [gene_to_int[g] for g in self.coords[self.Constants.GENE.value].values]
+        gene_array = np.array(gene_list)
+        decoded_image = gene_array.reshape(self.attrs['image_shape'])  # reverse linearization
+
+        # label the image and extract max intensity across each feature
+        label_image = label(decoded_image, connectivity=2)
+        intensities = self.max(dim=[Indices.CH, Indices.HYB])  # todo can use mean here too
```
Comment (dganguli) [inline starfish/intensity_table.py]: I think this supposed to be the magnitude of the pixel traces. In other words, for each pixel, it's the magnitude of the pixel trace.

File: starfish/pipeline/features/pixels/pixel_spot_detector.py [hunk 5]
```diff
@@ -34,132 +28,31 @@ def __init__(
...
-            new_index_name='barcode_index',
-            new_value_name='intensity',
-            melt_columns=range(num_bits)
+    def find(self, stack: ImageStack) \
+            -> Tuple[IntensityTable, List, np.ndarray, np.ndarray, np.ndarray]:
+        pixel_intensities = stack.to_pixel_intensities(crop=self.crop_size)
+        decoded_intensities = self.codebook.decode_euclidean(
+            pixel_intensities,
+            max_distance=self.distance_threshold,
+            min_intensity=self.magnitude_threshold
```
Comment (dganguli) [inline starfish/pipeline/features/pixels/pixel_spot_detector.py]: i'm not sure this is working as intended -- let's trace through the use of this parameter on a whiteboard to make certain.

File: starfish/pipeline/features/pixels/pixel_spot_detector.py (hunk 2 above)
Comment (ambrosejcarr) [reply starfish/pipeline/features/pixels/pixel_spot_detector.py]: I think this is a good idea. I also want to generalize some of this code so other spot detectors can use it. 

Used a namedtuple for now.

File: starfish/pipeline/features/pixels/pixel_spot_detector.py (hunk 1 above)
Comment (ambrosejcarr) [reply starfish/pipeline/features/pixels/pixel_spot_detector.py]: I can move the loading of the codebook into the PipelineComponent and remove this flexibility from the class. 👍 

It's a symptom of me implementing this before the paired PipelineComponent, thanks for catching.

File: starfish/pipeline/features/pixels/pixel_spot_detector.py [hunk 6]
```diff
@@ -1,173 +1,87 @@
...
+        pixel_intensities = IntensityTable.from_image_stack(
+            stack, crop_x=self.crop_x, crop_y=self.crop_y, crop_z=self.crop_z)
+        decoded_intensities = self.codebook.decode_euclidean(
+            pixel_intensities,
+            max_distance=self.distance_threshold,
+            min_intensity=self.magnitude_threshold
+        )
+        decoded_spots, image_decoding_results = decoded_intensities.combine_adjacent_features(
+            area_threshold=self.area_threshold,
+            assume_contiguous=True
```
Comment (dganguli) [inline starfish/pipeline/features/pixels/pixel_spot_detector.py]: still think we should nuke this --e.g., assume_contiguous is always true.

File: starfish/pipeline/features/pixels/__init__.py [hunk 7]
```diff
@@ -44,22 +46,22 @@ def _cli(cls, args, print_help=False):
...
+        intensities, decoded_props = instance.find(s)
 
         # TODO ambrosejcarr: this still needs to be added back.
         # if args.show:
         #     encoded_pixels.show(figsize=(10, 10))
 
-        path = os.path.join(args.output, 'pixels.geojson')
-        print(f"Writing | pixels geojson to: {path}")
-        pixel_attributes.save_geojson(path)
+        # path = os.path.join(args.output, 'pixels.geojson')
```
[the remark this answers is not in the document]
Comment (ambrosejcarr) [reply starfish/pipeline/features/pixels/__init__.py]: Left it as a reminder, but will add an explicit #TODO.

File: starfish/pipeline/features/pixels/combine_adjacent_features.py [hunk 8]
```diff
@@ -114,4 +117,7 @@ def combine_adjacent_features(
         attrs=attrs
     )
 
-    return intensity_table, props, label_image, decoded_image
+    ccdr = ConnectedComponentDecodingResult(props, label_image, decoded_image)
+    # import pdb; pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (ambrosejcarr) [reply starfish/pipeline/features/pixels/combine_adjacent_features.py]: 😭

File: starfish/pipeline/features/pixels/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ttung) [reply starfish/pipeline/features/pixels/__init__.py]: TODO: this pls.

File: starfish/pipeline/features/pixels/combine_adjacent_features.py [hunk 9]
```diff
@@ -1,4 +1,4 @@
-from typing import List, Tuple, Union
+from typing import List, Tuple, Union, NamedTuple
```
Comment (ttung) [inline starfish/pipeline/features/pixels/combine_adjacent_features.py]: sort pls.

File: starfish/pipeline/features/pixels/pixel_spot_detector.py [hunk 10]
```diff
@@ -1,173 +1,87 @@
...
+            spots whose codewords are more than this distance from an expected code are filtered
+            (default = 0.5176)
         magnitude_threshold : int
             spots with intensity less than this value are filtered (default 1)
         area_threshold : int
             spots with total area less than this value are filtered
-        crop_size : int
-            ??? # TODO ambrosejcarr what is this?
+        crop_x, crop_y, crop_z : int
+            number of pixels to crop from each of the x, y, and z axes of an ImageStack
```
Comment (ttung) [inline starfish/pipeline/features/pixels/pixel_spot_detector.py]: deceptive!

File: starfish/pipeline/features/pixels/pixel_spot_detector.py [hunk 11]
```diff
@@ -1,173 +1,87 @@
...
         self.magnitude_threshold = magnitude_threshold
         self.area_threshold = area_threshold
-        self.crop_size = crop_size
+        self.crop_x = crop_x
+        self.crop_y = crop_y
+        self.crop_z = crop_z
 
-        self.label_image: Optional[np.ndarray] = None
-        self.decoded_image: Optional[np.ndarray] = None
+    def find(self, stack: ImageStack) \
```
Comment (ttung) [inline starfish/pipeline/features/pixels/pixel_spot_detector.py]: does this not fit on one line?

File: starfish/pipeline/features/pixels/pixel_spot_detector.py (hunk 10 above)
Comment (ambrosejcarr) [reply starfish/pipeline/features/pixels/pixel_spot_detector.py]: Fixed.

File: starfish/pipeline/features/pixels/pixel_spot_detector.py (hunk 11 above)
Comment (ambrosejcarr) [reply starfish/pipeline/features/pixels/pixel_spot_detector.py]: Not on one 100 character line :)

**Pull request merged**